feat: `ExecutionTaggingIndexCache`

benesjan opened the pull request:
Fixes https://github.com/AztecProtocol/aztec-packages/issues/15753

In this PR I implement an ExecutionTaggingIndexCache which stores the used indexes during an execution and at the end we store them in the database. By storing that info once execution is done we'll be able to have indexes associated with a transaction hash. This is a necessary step towards improving of unconstrained tagging as described in [this post](https://forum.aztec.network/t/on-note-discovery-and-index-coordination/7165).

Given that once this PR is merged we will store the indexes in the database only for executions and not for simulations this PR should fix https://github.com/AztecProtocol/aztec-packages/issues/15753.

benesjan commented:
* **#19290** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19290?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19233** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19233?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19125** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19125?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19161** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19161?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19158** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19158?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19157** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19157?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#19030** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/19030?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17611** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17611?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17568** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17568?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17534** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17445** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17445?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17445?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#17451** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17451?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/noir-contracts/contracts/test/test_log_contract/src/main.nr:
```diff
@@ -75,4 +76,36 @@ pub contract TestLog {
     }
+
+    #[private]
+    fn emit_encrypted_events_nested(other: AztecAddress, num_nested_calls: u32) {
```
We don't yet have a convenient way how to test the oracles so I added this such that I can verify that the tags are unique. The nested calls allows me to check that we correctly pass the cache between calls.

nventuro replied:
You can test them in noir:

```noir
#[test]
fn jans_oracle_test() {
  let env = TestEnvironment::new();
  env.private_context(|context| {
  emit_event_in_private(
    ExampleEvent0 { value0: random_value_0, value1: random_value_1 },
      &mut context,
      other,
      MessageDelivery.UNCONSTRAINED_ONCHAIN,
      );
  });
}
```

The `#[event]` macro would not work so we'd need to create the struct manually. We already have a precedent for this in `MockNote` - you could place this as a mock event struct next to that.

nventuro replied:
(I'm not sure what it is you want to test however. This will definitely call the oracles, but txe does not currently use the logs in the block it creates, so if you want to search for those then this won't work. I created #17470)

benesjan replied:
[This comment](https://github.com/AztecProtocol/aztec-packages/pull/17445#discussion_r2406317656) applies here as well.

benesjan commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -751,10 +751,26 @@ export class PXE {
           nodeRPCCalls: contractFunctionSimulator?.getStats().nodeRPCCalls,
         });
+
+        // TODO(benesjan): What to do with this inefficiency? Could we just return the Tx directly from here?
+        // const txHash = (await txProvingResult.toTx()).getTxHash();
```
In a followup PR I will need to obtain a tx hash here which will most likely require some refactorings.

nventuro replied:
I don't understand, what's the inefficiency?

benesjan replied:
Sorry. Realized I was very messy in this PR.

I originally thought I will implement the association of index with a tx hash in this PR and left this comment here.

The future inefficiency is that we return from this function just a TxProvingResult but we will need to get a hold of tx hash, which is a costly operation.

Nuked the comment.

nventuro commented on yarn-project/end-to-end/src/e2e_event_logs.test.ts:
```diff
@@ -160,5 +160,60 @@ describe('Logs', () => {
+
+    // TODO: If we manage to encapsulate tagging related functionality we could most likely replace this test with
+    // a unit test.
+    it('produces unique tags for encrypted logs across nested calls and different transactions', async () => {
+      let tx1Tags: string[];
```
I'm not entirely sure what it is you're trying to test, that the tags are different because they have different indices and are emitted by different contracts? Not sure how nesting is relevant.

If that's what you're after, you can do something like

```noir
#[test]
fn events_get_unique_tags() {
  let env = TestEnvironment::new();
  env.private_context(|context| {
    emit_event(ctx, event);
    emit_event(ctx, event);
   assert(context.logs[0].get(0) != context.logs[1].get(0));
  });
}

#[test]
fn different_contracts_get_unique_tags() {
  let env = TestEnvironment::new();
  let log_a =  env.private_context_at(addr_a, |context| {
    emit_event(ctx, event);
    context.logs[0]
  });

  let lob_b = env.private_context_at(addr_b, |context| {
    emit_event(ctx, event);
    context.logs[0]
  });

  assert(log_a[0] != log_b[0]);
}

nventuro replied:
Alternatively if you were to put the tagging stuff outside of the private execution oracle it'd also be easier to unit test. These are more smoke tests, probably still useful.

benesjan replied:
Sorry, I was not clear enough here.

I wanted to test that the `ExecutionTaggingIndexCache` is correctly passed around when we nest. I don't think we have a way to test that out of e2e, right?

If not, will just improve the test description.

benesjan replied:
We discussed in the office where we agreed the current e2e test makes sense. Clarified the test purpose in https://github.com/AztecProtocol/aztec-packages/pull/17445/commits/69cc16196172dd0330a150aeb6ae17b9d3b99b8c

nventuro commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution_oracle.ts:
```diff
@@ -148,6 +151,13 @@ export class PrivateExecutionOracle extends UtilityExecutionOracle implements IP
   }
 
+  /**
+   * Return the tagging indexes incremented by this execution along with the directional app tagging secrets.
+   */
+  public getIndexedTaggingSecrets(): IndexedTaggingSecret[] {
+    return this.taggingIndexCache.getIndexedTaggingSecrets();
+  }
```
This solidified me not liking the 'indexed tagging secret' name - I just don't really know what it means. The 'tagging secret' part is particuarly annoying because it hides a lot of stuff and makes me have to think about the complexity of the tag derivation. Like, the tagging secret is not really indexed - that'd be the tag. What this is is the 'app siloed sender/recipient shared secret'. Perhaps we need to come up with a name for the thing that we combine with an index to get a tag, e.g. the 'pretag', 'tag seed', or whatever.

At any rate, I'd also change the name of this function, which is not very descriptive. Get which tagging secrets? I'd call it 'getUsedTags' or something nicer that conveys that these are the tags that this execution consumed and should be a) tracked by tx and b) never used again.

benesjan replied:
You have a point. The `IndexedTaggingSecret` name is indeed bad as the point of it is not indexing a secret (indexing in a sense of it being easily obtainable from a data store) but it is either used as a preimage of a tag or as a data container to send to the db to update the next index to be used for a given directional app tagging secret.

As to the name of the function it's quite confusing here because it's not the last used indexes but it's the indexes that are to be used in a tx. This is done because it makes the tagging data provider simpler as then we never need to return undefined from the `getNextIndexAsSender` function as by returning next we can just return 0 in the beginning:

<img width="901" height="213" alt="Image" src="https://github.com/user-attachments/assets/e53a279e-efe6-45af-ab3b-d345060affc0" />

But anyway I think it would be a good way to change this as handling undefined in 1 place is better then having all these function confusing.

I propose the following:
1. Rename `IndexedTaggingSecret` as `PreTag`- it's a preimage of a tag so I think this name is clear,
2. refactor all the functions to return last used tag instead of the next one as it's clearer,
3. rename this function to `getUsedPreTags`.

Doing all these refactor would explode the scope of this PR so I would do that in a followup one.

Sounds good?

nventuro commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution_oracle.ts:
```diff
@@ -198,15 +208,29 @@ export class PrivateExecutionOracle extends UtilityExecutionOracle implements IP
+        directionalAppTaggingSecret,
+        recipient,
+      });
+      await this.executionDataProvider.syncTaggedLogsAsSender(directionalAppTaggingSecret, this.contractAddress);
```
It's not obvious why we only sync if we didn't get an index. If `cache.getTaggingIndex` was called something else and with some light restructuring this might be more obvious.

```ts
function getIndexToUseForSecret(secret) {
  const lastUsedIndexInTx =  this.taggingIndexCache.getLastusedIndex(directionalAppTaggingSecret);

  if (lastUsedIndexInTx) {
    return lastUsedIndexInTx + 1;
  } else {
    // this is a tagging secret we've not yet used in this tx, so first sync our store to make sure its indices
    // are up to date. we do this here because this store is not synced as part of the global sync because
    // that'd be wasteful as most tagging secrets are not used in each tx
    await this.executionDataProvider.syncTaggedLogsAsSender(secret,  this.contractAddress);
    const lastUsedIndex = await this.executionDataProvider.getTaggingSecretsIndexAsSender(secret);
    return lastUsedIndex + 1;
  }
}

function foo() {
  const index = getIndexToUseForSecret(secret);
  this.cache.storeUsedIndex(secret, index);
```

benesjan replied:
That's indeed much better. As mentioned [here](https://github.com/AztecProtocol/aztec-packages/pull/17445#discussion_r2405913527) will do the change of returning last used index instead of next index in a followup PR.

Included your comment in [this commit](https://github.com/AztecProtocol/aztec-packages/pull/17445/commits/182377add842d0d83f69667d1a60dd2f5a0ab172) and put there a note to make the function look as you proposed.

nventuro commented on yarn-project/pxe/src/contract_function_simulator/execution_tagging_index_cache.ts:
```diff
@@ -0,0 +1,26 @@
+
+  public setTaggingIndex(directionalAppTaggingSecret: Fr, index: number) {
+    this.taggingIndexMap.set(directionalAppTaggingSecret.toString(), index);
+  }
```
It'd be good be a reasonable sanity check to assert that if the map contains this secret already, the index is the one immediately below this one (i.e. `index  - 1`). We recently had a bug that would've been detected by such a check (https://github.com/AztecProtocol/aztec-packages/pull/17317).

benesjan replied:
Smort. Implemented in https://github.com/AztecProtocol/aztec-packages/pull/17445/commits/fd39b63ec683e3823c1e51e50e42dd173205d71d

Thanks

benesjan commented:
@nventuro I had to rebase this since I heavily modified the PR down the stack but all the changes I did since your last review are just in these commits so feel free to just check those:

<img width="970" height="251" alt="image" src="https://github.com/user-attachments/assets/b8e946cb-527b-4cfe-bc06-f4eb97314bb5" />

benesjan committed:
feat: `ExecutionTaggingIndexCache`

Fixes https://github.com/AztecProtocol/aztec-packages/issues/15753

In this PR I implement an ExecutionTaggingIndexCache which stores the used indexes during an execution and at the end we store them in the database. By storing that info once execution is done we'll be able to have indexes associated with a transaction hash. This is a necessary step towards improving of unconstrained tagging as described in [this post](https://forum.aztec.network/t/on-note-discovery-and-index-coordination/7165).

Given that once this PR is merged we will store the indexes in the database only for executions and not for simulations this PR should fix https://github.com/AztecProtocol/aztec-packages/issues/15753.
